### PR TITLE
UCP/API: client-server connection establishment

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
## What
This PR adds new UCP client-server connection establishment flow. The main difference from origin is possibility to handle connection requests on listener created on one worker but create an endpoint ("accept") on other worker. Also added ucp_listener_reject function to avoid memory leaks

## Why ?
The existing limitation may be critical for porting socket-based apps to UCX.

(https://github.com/evgeny-leksikov/ucx/pull/3)

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7600/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7636/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7644/ for details (Mellanox internal link).

File: src/ucp/api/ucp_def.h [hunk 1]
```diff
@@ -376,18 +375,20 @@ typedef void
...
- * thread safe with respect to the worker it is invoked on. Also it is an
- * user's responsibility to avoid possible dead lock accessing different worker.
- * If the callback is called from different threads, this callback needs thread
- * safety support.
+ * the listener. Setting this type of handler allows creating an endpoint on
+ * any other worker and not limited to the worker on which the listener was
+ * created.
+ * @note
+ * - Other than communication progress routines, it is allowed to call
+ *   other communication routines from the callback in the struct.
```
Comment (yosefe) [inline src/ucp/api/ucp_def.h]: all other

File: src/ucp/api/ucp_def.h [hunk 2]
```diff
@@ -376,18 +375,20 @@ typedef void
...
+ * - The callback is thread safe with respect to the worker it is invoked on.
+ * - It is an user's responsibility to avoid possible dead lock accessing
```
Comment (yosefe) [inline src/ucp/api/ucp_def.h]: It is the user's
possible -> potential

File: src/ucp/api/ucp_def.h [hunk 3]
```diff
@@ -376,18 +375,20 @@ typedef void
...
+ *   different worker.
+ * - If the callback is called from different threads, this callback needs
```
Comment (yosefe) [inline src/ucp/api/ucp_def.h]: the callback is always called from the context of ucp_worker_progress, right? so don't understand why this note is needed..

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7665/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@shamisp @gmegan pls take a look

**Comment by yosefe** [comment]
@shamisp good to go?

**Comment by shamisp** [comment]
👍

**Comment by evgeny-leksikov** [comment]
What about https://github.com/openucx/ucx/pull/2839? I can't add UCP implementation before UCT

**Pull request merged**